attest/eventlog.go

```diff
@@ -241,46 +249,88 @@ func extend(pcr, replay []byte, e rawEvent) ([]byte, Event, error) {
 			hash.Write(b)
 		}
 		hash.Write(digest)
-		return hash.Sum(nil), Event{e.index, e.typ, e.data, digest}, nil
+		return hash.Sum(nil), digest, nil
 	}
-	return nil, Event{}, fmt.Errorf("no event digest matches pcr length: %d", len(pcr))
+	return nil, nil, fmt.Errorf("no event digest matches pcr length: %d", len(pcr))
 }
 
-func replayEvents(rawEvents []rawEvent, pcrs []PCR) ([]Event, error) {
-	events := []Event{}
-	replay := map[int][]byte{}
-	pcrByIndex := map[int][]byte{}
-	for _, pcr := range pcrs {
-		pcrByIndex[pcr.Index] = pcr.Digest
-	}
+// replayPCR replays the event log for a specific PCR, using pcr and
+// event digests with the algorithm in pcr. An error is returned if the
+// replayed values do not match the final PCR digest, or any event tagged
+// with that PCR does not posess an event digest with the specified algorithm.
+func replayPCR(rawEvents []rawEvent, pcr PCR) ([]Event, error) {
+	var (
+		replay    []byte
+		hadEvents bool
+	)
 
+	var outEvents []Event
 	for i, e := range rawEvents {
-		pcrValue, ok := pcrByIndex[e.index]
-		if !ok {
-			// Ignore events for PCRs that weren't included in the quote.
+		if e.index != pcr.Index {
 			continue
 		}
-		replayValue, event, err := extend(pcrValue, replay[e.index], e)
+		hadEvents = true
+
+		replayValue, digest, err := extend(pcr.Digest, replay, e)
 		if err != nil {
-			return nil, fmt.Errorf("replaying event %d: %v", i, err)
+			return nil, fmt.Errorf("replaying event[%d] in PCR[%d]: %v", i, pcr.Index, err)
 		}
-		replay[e.index] = replayValue
-		events = append(events, event)
+		replay = replayValue
+		outEvents = append(outEvents, Event{sequence: e.sequence, Data: e.data, Digest: digest, Index: pcr.Index, Type: e.typ})
+	}
+
+	if hadEvents && !bytes.Equal(replay, pcr.Digest) {
+		return nil, fmt.Errorf("PCR[%d] failed to replay using crypto.Hash(%v)", pcr.Index, pcr.DigestAlg)
 	}
+	return outEvents, nil
+}
+
+type pcrReplayResult struct {
+	events []Event
+	err    error
+}
+
+func replayEvents(rawEvents []rawEvent, pcrs []PCR) ([]Event, error) {
+	var (
+		invalidReplays []int
+		verifiedEvents []Event
+		allPCRReplays  = map[int][]pcrReplayResult{}
+	)
 
-	var invalidReplays []int
-	for i, value := range replay {
-		if !bytes.Equal(value, pcrByIndex[i]) {
-			invalidReplays = append(invalidReplays, i)
+	// Replay the event log for every PCR and digest algorithm combination.
+	for _, pcr := range pcrs {
+		events, replayErr := replayPCR(rawEvents, pcr)
```

Comment:
instead of an error, maybe

```
events, ok := replayPCR(rawEvents, pcr)
if ok {
    allPCRReplays[pcr.Index] = append(allPCRReplays[pcr.Index], pcrReplayResult{events})
}
```

Reply:
In that case, `allPCRReplays` would only contain successful replays, so while we would omit returning events from unverified PCRs we would not error if both failed to replay?